### PR TITLE
Remove some n^2 scaling in interchange

### PR DESCRIPTION
**Pull request by benclifford**
Prior to this PR, the interchange inspects every worker registration for every task message received.

This means when allocating n tasks to n workers, this is O(n^2)

In practice with imsim in production this was producing noticeable slowdowns.

This PR changes the behaviour so that a set of interesting workers is maintained, and only interesting workers are inspected when allocating a task. Interesting workers are the workers which have available capacity.

In the imsim work which motivated this PR, there are on the order of 2000 workers, and a larger number of long running (many hour) jobs. Testing and thinking-about should happen for different workloads (for exmaple, the case where workers are starved of work, and when jobs are very short.

### CODE REVIEW

**Comment by benclifford** [comment]
todo: remove some of the debug log lines in here

File: parsl/executors/high_throughput/interchange.py [hunk 1]
```diff
@@ -378,15 +389,28 @@ def start(self, poll_period=1):
...
 
             # If we had received any requests, check if there are tasks that could be passed
-            logger.debug("Managers: {}".format(self._ready_manager_queue))
-            if self._ready_manager_queue:
-                shuffled_managers = list(self._ready_manager_queue.keys())
+
+            # even rendering the list here is maybe expensive, even if not written out
+            # logger.debug("Managers: {}".format(self._ready_manager_queue))
+            logger.debug("Managers count: {}".format(len(self._ready_manager_queue)))
+            logger.debug("Interesting managers count: {}".format(len(interesting_managers)))
```
Comment (benclifford) [inline parsl/executors/high_throughput/interchange.py]: join these into one line

File: parsl/executors/high_throughput/interchange.py [hunk 2]
```diff
@@ -378,15 +389,28 @@ def start(self, poll_period=1):
...
+                logger.debug("[MAIN] leaving task_outgoing section")
 
             # If we had received any requests, check if there are tasks that could be passed
-            logger.debug("Managers: {}".format(self._ready_manager_queue))
-            if self._ready_manager_queue:
-                shuffled_managers = list(self._ready_manager_queue.keys())
+
+            # even rendering the list here is maybe expensive, even if not written out
+            # logger.debug("Managers: {}".format(self._ready_manager_queue))
+            logger.debug("Managers count: {}".format(len(self._ready_manager_queue)))
```
Comment (yadudoc) [inline parsl/executors/high_throughput/interchange.py]: We could merge 399 and 400.

File: parsl/executors/high_throughput/interchange.py [hunk 3]
```diff
@@ -378,15 +389,28 @@ def start(self, poll_period=1):
...
+            logger.debug("Interesting managers count: {}".format(len(interesting_managers)))
+
+            # some managers will still be interesting after this pass... accumulate them here
+            if interesting_managers and not self.pending_task_queue.empty():
+                logger.debug("[MAIN] entering _ready_manager_queue section")
+                logger.debug("[MAIN]   copying manager list to shuffle")
+                shuffled_managers = list(interesting_managers)
+                logger.debug("[MAIN]   shuffling list")
```
Comment (benclifford) [inline parsl/executors/high_throughput/interchange.py]: remove

File: parsl/executors/high_throughput/interchange.py [hunk 4]
```diff
@@ -378,15 +389,28 @@ def start(self, poll_period=1):
...
-                shuffled_managers = list(self._ready_manager_queue.keys())
+
+            # even rendering the list here is maybe expensive, even if not written out
+            # logger.debug("Managers: {}".format(self._ready_manager_queue))
+            logger.debug("Managers count: {}".format(len(self._ready_manager_queue)))
+            logger.debug("Interesting managers count: {}".format(len(interesting_managers)))
+
+            # some managers will still be interesting after this pass... accumulate them here
+            if interesting_managers and not self.pending_task_queue.empty():
+                logger.debug("[MAIN] entering _ready_manager_queue section")
```
Comment (yadudoc) [inline parsl/executors/high_throughput/interchange.py]: We should remove these timing log lines

File: parsl/executors/high_throughput/interchange.py [hunk 5]
```diff
@@ -378,15 +389,28 @@ def start(self, poll_period=1):
...
+                logger.debug("[MAIN]   copying manager list to shuffle")
+                shuffled_managers = list(interesting_managers)
+                logger.debug("[MAIN]   shuffling list")
                 random.shuffle(shuffled_managers)
-                logger.debug("Shuffled: {}".format(shuffled_managers))
+                logger.debug("[MAIN]   shuffled list")
```
Comment (benclifford) [inline parsl/executors/high_throughput/interchange.py]: remove this log

File: parsl/executors/high_throughput/interchange.py [hunk 6]
```diff
@@ -378,15 +389,28 @@ def start(self, poll_period=1):
...
+                # logger.debug("Shuffled : {}".format(shuffled_managers))
                 # for manager in self._ready_manager_queue:
-                for manager in shuffled_managers:
+                while shuffled_managers and not self.pending_task_queue.empty():  # cf. the if statement above...
```
Comment (yadudoc) [inline parsl/executors/high_throughput/interchange.py]: :+1:

**Comment by benclifford** [comment]
@yadudoc the requested changes should be done now

**Pull request merged**